File: backend/schemas/shops.py
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field


# ============ 请求模型 ============

class MenuItemCreateRequest(BaseModel):
    """创建菜单项请求"""
    name: str = Field(min_length=1, max_length=100, description="菜品名称")
    price: Optional[float] = Field(default=None, ge=0, description="价格")
    description: Optional[str] = Field(default=None, max_length=500, description="菜品描述")


class MenuItemAddRequest(BaseModel):
    """添加菜单项请求"""
    name: str = Field(..., min_length=1, max_length=100, description="菜品名称")
    price: float = Field(..., gt=0, description="价格（必填）")
    description: Optional[str] = Field(default=None, max_length=500, description="菜品描述（可选）")


class ShopCreate(BaseModel):
    """创建店铺请求"""
    name: str = Field(min_length=1, max_length=100, description="店铺名称")
    dict_data_ids: List[int] = Field(
        ..., min_length=2,
        description="字典数据ID列表，至少包含一个品类ID和一个区域ID"
    )
    menu_items: Optional[List[MenuItemCreateRequest]] = Field(default=None, description="初始菜单项（可选）")


class ShopUpdate(BaseModel):
    """更新店铺请求（管理员）"""
    name: Optional[str] = Field(default=None, min_length=1, max_length=100, description="店铺名称")
    is_active: Optional[bool] = Field(default=None, description="是否启用（软删除）")
    dict_data_ids: Optional[List[int]] = Field(
        default=None,
        description="字典数据ID列表（整体替换店铺标签，传空列表则清空标签）"
    )


class RatingCreate(BaseModel):
    """评分请求"""
    score: int = Field(ge=1, le=5, description="评分值（1-5）")


# ============ 响应模型 ============

class RatingDistribution(BaseModel):
    """评分分布统计"""
    star_1: int = Field(default=0, description="1星评分人数")
    star_2: int = Field(default=0, description="2星评分人数")
    star_3: int = Field(default=0, description="3星评分人数")
    star_4: int = Field(default=0, description="4星评分人数")
    star_5: int = Field(default=0, description="5星评分人数")
    total: int = Field(default=0, description="总评分人数")

    class Config:
        from_attributes = True


class MenuItemResponse(BaseModel):
    """菜单项响应"""
    id: int = Field(description="菜单项ID")
    shop_id: int = Field(description="所属店铺ID")
    name: str = Field(description="菜品名称")
    price: Optional[float] = Field(default=None, description="价格")
    description: Optional[str] = Field(default=None, description="菜品描述")
    created_at: datetime = Field(description="创建时间")

    class Config:
        from_attributes = True


class RatingResponse(BaseModel):
    """评分响应"""
    id: int = Field(description="评分ID")
    user_id: int = Field(description="评分用户ID")
    shop_id: int = Field(description="店铺ID")
    score: int = Field(description="评分值")
    created_at: datetime = Field(description="评分时间")
    updated_at: datetime = Field(description="更新时间")

    class Config:
        from_attributes = True


class DictDataSimpleResponse(BaseModel):
    """字典数据简略响应（店铺标签用）"""
    id: int = Field(description="字典数据ID")
    name: str = Field(description="标签名称")
    extra: Optional[dict] = Field(default=None, description="扩展信息")

    class Config:
        from_attributes = True


class ImageBriefResponse(BaseModel):
    """图片简略响应（店铺详情/列表用）"""
    id: int = Field(description="图片ID")
    url: str = Field(description="图片URL")

    class Config:
        from_attributes = True


class ShopResponse(BaseModel):
    """店铺详情响应"""
    id: int = Field(description="店铺ID")
    name: str = Field(description="店铺名称")
    view_count: int = Field(description="浏览量")
    favorite_count: int = Field(description="收藏数")
    comment_count: int = Field(description="评论数")
    average_rating: float = Field(default=0.0, description="平均评分")
    rating_distribution: Optional[RatingDistribution] = Field(default=None, description="评分分布统计")
    aliases: Optional[List[str]] = Field(default=None, description="别名列表")
    merged_into_id: Optional[int] = Field(default=None, description="合并后店铺ID")
    is_banned: bool = Field(default=False, description="是否被封禁")
    dict_data: Optional[List[DictDataSimpleResponse]] = Field(default=None, description="关联的字典标签")
    menu_items: Optional[List[MenuItemResponse]] = Field(default=None, description="菜单列表")
    images: Optional[List[ImageBriefResponse]] = Field(default=None, description="图片列表")
    is_favorited: bool = Field(default=False, description="当前用户是否已收藏")
    user_rating: Optional[RatingResponse] = Field(default=None, description="当前用户的评分")
    created_at: datetime = Field(description="创建时间")
    updated_at: datetime = Field(description="更新时间")

    class Config:
        from_attributes = True


class ShopListItem(BaseModel):
    """店铺列表项"""
    id: int = Field(description="店铺ID")
    name: str = Field(description="店铺名称")
    average_rating: float = Field(default=0.0, description="平均评分")
    view_count: int = Field(description="浏览量")
    favorite_count: int = Field(description="收藏数")
    comment_count: int = Field(description="评论数")
    cover_image: Optional[str] = Field(default=None, description="封面图片URL")
    dict_data: Optional[List[DictDataSimpleResponse]] = Field(default=None, description="字典标签")
    is_favorited: bool = Field(default=False, description="当前用户是否已收藏")
    created_at: datetime = Field(description="创建时间")

    class Config:
        from_attributes = True

File: frontend/src/api/shops.ts
import apiClient from './client';
import type { ApiResponse } from '../types/common';

// ============ 类型定义 ============

export interface ShopCardData {
  id: number;
  name: string;
  average_rating: number;
  view_count: number;
  favorite_count: number;
  comment_count: number;
  cover_image: string | null;
  dict_data: { id: number; name: string }[];
  is_favorited: boolean;
  created_at: string;
}

export interface ShopDetail extends ShopCardData {
  rating_distribution: { star_1: number; star_2: number; star_3: number; star_4: number; star_5: number; total: number };
  aliases: string[] | null;
  merged_into_id: number | null;
  is_banned: boolean;
  menu_items: MenuItem[];
  images: { id: number; url: string }[];
  user_rating: { score: number } | null;
  updated_at: string;
}

export interface MenuItem {
  id: number;
  shop_id: number;
  name: string;
  price: number | null;
  description: string | null;
  created_at: string;
}

export interface ShopFilter {
  keyword?: string;
  category_ids?: number[];
  district_ids?: number[];
  min_rating?: number;
  sort_by?: string;
  sort_order?: string;
  page: number;
  page_size: number;
}

export interface CreateShopData {
  name: string;
  dict_data_ids: number[];
  menu_items?: { name: string; price?: number; description?: string }[];
}

export interface UpdateShopData {
  name?: string;
  dict_data_ids?: number[];
  is_active?: boolean;
}

export interface DictItem {
  id: number;
  dict_type_id: number;
  name: string;
}

export interface DictType {
  id: number;
  name: string;
  target_table: string;
}

// ============ 店铺接口 ============

export async function fetchShops(filter: ShopFilter): Promise<{ items: ShopCardData[]; total: number }> {
  const params: Record<string, any> = {
    keyword: filter.keyword || undefined,
    category_ids: filter.category_ids?.length ? filter.category_ids.join(',') : undefined,
    district_ids: filter.district_ids?.length ? filter.district_ids.join(',') : undefined,
    min_rating: filter.min_rating,
    sort_by: filter.sort_by || 'favorite_count',
    sort_order: filter.sort_order || 'desc',
    page: filter.page,
    page_size: filter.page_size || 20,
  };
  const res = await apiClient.get<ApiResponse<{ items: ShopCardData[]; total: number }>>('/shops', { params });
  return res.data.data;
}

export async function fetchShopDetail(shopId: number): Promise<ShopDetail> {
  const res = await apiClient.get<ApiResponse<ShopDetail>>(`/shops/${shopId}`);
  return res.data.data;
}

export async function createShopApi(data: CreateShopData): Promise<ShopDetail> {
  const res = await apiClient.post<ApiResponse<ShopDetail>>('/shops', data);
  return res.data.data;
}

export async function updateShopApi(shopId: number, data: UpdateShopData): Promise<ShopDetail> {
  const res = await apiClient.put<ApiResponse<ShopDetail>>(`/shops/${shopId}`, data);
  return res.data.data;
}

export async function deleteShopApi(shopId: number): Promise<void> {
  await apiClient.delete(`/shops/${shopId}`);
}

// ============ 评分接口 ============

export async function submitRating(shopId: number, score: number): Promise<{ average_rating: number; rating_distribution: any }> {
  const res = await apiClient.post<ApiResponse<{ average_rating: number; rating_distribution: any }>>(`/shops/${shopId}/rating`, { score });
  return res.data.data;
}

// ============ 菜单接口 ============

export async function fetchMenu(shopId: number): Promise<MenuItem[]> {
  const res = await apiClient.get<ApiResponse<MenuItem[]>>(`/shops/${shopId}/menu`);
  return res.data.data;
}

export async function addMenuItem(shopId: number, name: string, price?: number, description?: string): Promise<MenuItem> {
  const res = await apiClient.post<ApiResponse<MenuItem>>(`/shops/${shopId}/menu`, null, {
    params: { name, price, description },
  });
  return res.data.data;
}

// ============ 字典接口 ============

export async function fetchDictTypes(): Promise<{ id: number; name: string }[]> {
  const res = await apiClient.get('/dict/types');
  return (res.data as any).data;
}


